### PR TITLE
 Add achievement on every five sections submitted

### LINKED ISSUE #677: Add achievement on every five sections submitted
When submitting feedback, check total count of sections, divide them by 5, and check current status of the achievement via crew api. If the numbers don't match, add an achievement.

### PR DESCRIPTION
**Pull request by nyet-ty**
Resolve #677

## PR includes

-   [x] Feature

### CODE REVIEW

File: src/config.ts [hunk 1]
```diff
@@ -66,6 +66,8 @@ export default {
     crew: {
         url: process.env.NEXT_PUBLIC_CREW_URL,
         apiToken: process.env.CREW_API_TOKEN,
+        sectionAchiementId: process.env.CREW_SECTIONS_ACHIEVEMENT,
```
Comment (tsumo) [inline src/config.ts:69]: achieVement

File: src/modules/crewTypes.ts [hunk 2]
```diff
@@ -78,6 +78,7 @@ export interface Group {
     id: string;
     name: string;
     description?: string;
+    [key: string]: unknown;
```
Comment (tsumo) [inline src/modules/crewTypes.ts:81]: ?

File: src/modules/sectionMethods.ts [hunk 3]
```diff
@@ -202,6 +205,30 @@ const findAllInterviewerSections = async (
...
+    if (amountOfInterviewerCompletedSections % 5 !== 0) return;
+
+    const { email } = await userMethods.find(interviewerId);
+    const crewUser = await crewMethods.getUserInfo(email);
+
+    const sectionAchievementCount =
+        crewUser.achievements.find((a) => a.achievement.id === config.crew.sectionAchiementId)?.count || 0;
+    const amount = amountOfInterviewerCompletedSections / 5 - sectionAchievementCount;
+
+    amount > 0 &&
```
Comment (tsumo) [inline src/modules/sectionMethods.ts:223]: Use regular `if`, `&&` and `await` look bad together

File: src/modules/sectionMethods.ts [hunk 4]
```diff
@@ -229,14 +256,18 @@ const update = async (data: UpdateSection): Promise<Section> => {
...
     const updatedSection = await prisma.section.update({
         data: updateData,
         where: { id: sectionId },
         include: { sectionType: true },
     });
 
+    if (sendHrMail) {
+        await notifyHR(sectionId, data);
+
+        await giveSectionAchievement(interviewerId);
```
Comment (tsumo) [inline src/modules/sectionMethods.ts:268]: why this is under sendHrMail flag?

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/modules/sectionMethods.ts (hunk 4 above)
Comment (nyet-ty) [reply src/modules/sectionMethods.ts:268]: This flag became more universal, I'll rename it

File: src/modules/sectionMethods.ts [hunk 5]
```diff
@@ -202,9 +205,43 @@ const findAllInterviewerSections = async (
...
+    });
+
+    if (amountOfInterviewerCompletedSections % 5 !== 0) return;
+
+    const { email } = await userMethods.find(interviewerId);
+    const crewUser = await crewMethods.getUserInfo(email);
+
+    const sectionAchievementCount =
+        crewUser.achievements.find((a) => a.achievement.id === config.crew.sectionAchivementId)?.count || 0;
+    const amount = amountOfInterviewerCompletedSections / 5 - sectionAchievementCount;
```
Comment (9teen90nine) [inline src/modules/sectionMethods.ts:222]: 5 should be a const in config probably

File: src/modules/sectionMethods.ts [hunk 6]
```diff
@@ -202,9 +205,43 @@ const findAllInterviewerSections = async (
...
 };
 
+const giveSectionAchievement = async (interviewerId: number) => {
+    if (!config.crew.techUserEmail || !config.crew.sectionAchivementId) return;
+
+    const amountOfInterviewerCompletedSections = await prisma.section.count({
+        where: { interviewerId, feedback: { not: null } },
+    });
+
+    if (amountOfInterviewerCompletedSections % 5 !== 0) return;
```
Comment (9teen90nine) [inline src/modules/sectionMethods.ts:215]: here also

File: src/modules/crewMethods.ts [hunk 7]
```diff
@@ -67,9 +78,52 @@ export const crewMethods = {
...
+
+    giveAchievement: async (data: GiveAchievement) => {
+        const { url, apiToken } = checkConfig();
+        const response = await fetch(`${url}/api/rest/achievements/give`, {
+            method: 'POST',
+            headers: {
+                authorization: apiToken,
+                'Content-Type': 'application/json',
+            },
+            body: JSON.stringify(data),
```
Comment (9teen90nine) [inline src/modules/crewMethods.ts:125]: don't you want to create a wrapper for these fetches? it's a copypaste

**Review by 9teen90nine — CHANGES_REQUESTED** [review]

File: src/modules/crewMethods.ts (hunk 7 above)
Comment (nyet-ty) [reply src/modules/crewMethods.ts:125]: not really: all of these fetches have `authorization: apiToken,` but almost half of them are GET, so they don't have content type and body

Comment (nyet-ty) [reply src/modules/crewMethods.ts:125]: or maybe wrappers for get and wrapper for post apart?

File: src/modules/sectionMethods.ts [hunk 8]
```diff
@@ -202,9 +205,44 @@ const findAllInterviewerSections = async (
...
+    const crewUser = await crewMethods.getUserInfo(email);
+
+    const sectionAchievementCount =
+        crewUser.achievements.find((a) => a.achievement.id === config.crew.sectionAchivementId)?.count || 0;
+    const amount =
+        amountOfInterviewerCompletedSections / config.crew.sectionAmountForAchievement - sectionAchievementCount;
+
+    if (amount <= 0) return;
+
+    await crewMethods.giveAchievement({
```
Comment (tsumo) [inline src/modules/sectionMethods.ts:227]: We can also create section history record for giving achievement

File: src/components/HistoryRecord/HistoryRecord.tsx [hunk 9]
```diff
@@ -136,11 +136,18 @@ const SectionSetFeedback = ({ event }: EventProps) => {
     );
 };
 
+const AchievementsForSections = ({ event }: EventProps) => (
+    <Text size="xs" weight="thin">
+        {event.after}
```
Comment (tsumo) [inline src/components/HistoryRecord/HistoryRecord.tsx:141]: History record should keep only minimal data, not full messages

**Review by tsumo — CHANGES_REQUESTED** [review]

File: src/config.ts [hunk 10]
```diff
@@ -66,6 +66,9 @@ export default {
     crew: {
         url: process.env.NEXT_PUBLIC_CREW_URL,
         apiToken: process.env.CREW_API_TOKEN,
+        sectionAchiementId: process.env.CREW_SECTIONS_ACHIEVEMENT || 'ttrSaFjsA-9Gy-unh4zK2',
```
Comment (tsumo) [inline src/config.ts:69]: Probably too specific to store in repository

File: src/config.ts [hunk 11]
```diff
@@ -66,6 +66,9 @@ export default {
...
+        techUserEmail: process.env.CREW_TECHNICAL_USER_EMAIL || 'admin@taskany.org',
```
Comment (tsumo) [inline src/config.ts:70]: not even a real email

**Review by tsumo — CHANGES_REQUESTED** [review]

**Review by tsumo — APPROVED** [review]

**Review by asabotovich — APPROVED** [review]

**Review by 9teen90nine — APPROVED** [review]

**Pull request merged**